esp/services/ws_packageprocess/ws_packageprocessService.cpp

```diff
@@ -735,6 +812,87 @@ bool CWsPackageProcessEx::onAddPackage(IEspContext &context, IEspAddPackageReque
     return true;
 }
 
+#define PACKAGEMAP_PATH_SEP_CHAR '/'
+
+bool nextPmPathNode(const char *&path, StringBuffer &node)
+{
+    if (*path==PACKAGEMAP_PATH_SEP_CHAR)
+        path++;
+    while (*path && *path!=PACKAGEMAP_PATH_SEP_CHAR)
+        node.append(*path++);
+    return (*path && *++path);
+}
+
+bool splitPMPath(const char *path, StringBuffer &netAddress, StringBuffer &target, StringBuffer *pmid)
+{
+    if (!path || !*path)
+        return false;
+    if (*path==PACKAGEMAP_PATH_SEP_CHAR && path[1]==PACKAGEMAP_PATH_SEP_CHAR)
+    {
+        path+=2;
+        if (!nextPmPathNode(path, netAddress))
+            return false;
+    }
+    if (!nextPmPathNode(path, target))
+        return (pmid==NULL);
+    if (!pmid)
+        return false;
+    if (nextPmPathNode(path, *pmid))
+        return false; //query path too deep
+    return true;
+}
+
+
+bool CWsPackageProcessEx::onCopyPackageMap(IEspContext &context, IEspCopyPackageMapRequest &req, IEspCopyPackageMapResponse &resp)
+{
+    PackageMapUpdater updater;
+    updater.setFlag(PKGADD_MAP_CREATE);
+    updater.setFlag(PKGADD_MAP_ACTIVATE, req.getActivate());
+    updater.setFlag(PKGADD_MAP_REPLACE, req.getReplacePackageMap());
+    updater.setFlag(PKGADD_PRELOAD_ALL, req.getPreloadAllPackages());
+    updater.setFlag(PKGADD_ALLOW_FOREIGN, true);
+
+    updater.setProcess(req.getProcess());
+    updater.setUser(context.queryUserId(), context.queryPassword(), &context);
+    updater.setDerivedDfsLocation(req.getDaliIp(), req.getSourceProcess());
+
+    unsigned updateFlags = 0;
+    if (req.getReplacePackageMap())
+        updateFlags |= DALI_UPDATEF_PACKAGEMAP;
+    if (req.getUpdateCloneFrom())
+        updateFlags |= DALI_UPDATEF_CLONE_FROM;
+    if (req.getUpdateSuperFiles())
+        updateFlags |= DALI_UPDATEF_SUPERFILES;
+    if (req.getAppendCluster())
+        updateFlags |= DALI_UPDATEF_APPEND_CLUSTER;
+
+    StringBuffer srcAddress, srcTarget, srcPMID;
+    if (!splitPMPath(req.getSourcePath(), srcAddress, srcTarget, &srcPMID))
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid source PackageMap path");
+
+    if (req.getPMID() && *req.getPMID())
```

Comment:
cons char * requestedPMID = req.getPMID() and avoid calling req.getPMID a couple of times?